### PR TITLE
Feature powf, exp, log10, log2 log, ln, trunc, fract, copysign support + unit tests for each

### PR DESCRIPTION
**Pull request by Cazadorro**
There are about a million different implementations of each, but for exp and ln, you basically end up with a splitting based upon 2^n splitting and Horner schemes, or generated polynomials as the fastest methods.  You can tweak the iterations but overall there's really not much more you can do to make things faster, you generally end up with 4 -> 5 multiply adds on top of some bit-twiddling and a few non-iterative additions. 

Powf is pretty much always going to be based upon x^logx(b) scheme, and because of this rely heavily on both your exp/expx and ln/logx being reasonably accurate, as powf will inherit both functions error, if you skimp too much on accuracy you get extremely poor results.  That said, *relative* accuracy was pretty high in all of these tests, with powf being 0.002 *relative* (so not absolute like some of the tests present) in every instance tested (though not exhaustive). 

There are also some checks made for inputs of 1.0 and 0.0 in some functions, to provide exact output in those scenarios, I wasn't sure if they were necessary, but error is horrendous at the edges of the polynomial approximations, where 1.0 and 0.0 almost always were, resulting in some pretty bad error (ie I remember something like ln(1.0) == 0.0006... at some point in testing). 

The fract and trunc functions should work in all normal scenarios, but its possible that you would want versions which don't care about the sign of what is being extracted.  In such cases it might make more sense to have separate functions, but I wasn't sure if that was in the scope of "micromath" in general. 

This is my first time writing anything substantial in rust, and I'm not sure I've followed proper idioms, and I'm willing to make massive modifications if need be.

### CODE REVIEW

**Comment by tarcieri** [comment]
Nice! Will review this when I have some time.

File: src/f32ext/fract.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    let fractional_normalized: u32 =
+        fractional_part.overflowing_shl(exponent_shift).0 & utils::MANTISSA_MASK;
+
+    let new_exponent_bits = (utils::EXPONENT_BIAS - (exponent_shift))
+        .overflowing_shl(23_u32)
+        .0;
+    copysign::copysign(f32::from_bits(fractional_normalized | new_exponent_bits), x)
+}
+
+// don't want to delete maybe this is the better way?
```
Comment (tarcieri) [inline src/f32ext/fract.rs]: What is this specifically?

**Comment by tarcieri** [comment]
Overall it looks good, in the spirit of this library, and I don't see anything I'd consider particularly non-idiomatic.

I guess as a general note you have a lot of comments asking questions which seem to fall into a couple categories: things I'd consider TODOs, and rationales for the current implementation but phrased in the form of a question.

I think it'd be good to either label TODOs as TODOs or change the other questions to affirmations. Will try to leave some advice on those line by line.

File: src/f32ext/fract.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    let exponent_clamped = i32::max(exponent, 0_i32) as u32;
+
+    // find the part of the fraction that would be left over
+    let fractional_part: u32 = x_bits.overflowing_shl(exponent_clamped).0 & utils::MANTISSA_MASK;
+    // if there isn't a fraction we can just return 0
+    if fractional_part == 0_u32 {
+        // most people don't actually care about -0.0, so would it be better to just not copysign?
+        return copysign::copysign(0.0_f32, x);
+    }
+    //alternatively use -1.0? subtraction would surely be more costly though right?
```
Comment (tarcieri) [inline src/f32ext/fract.rs]: This is an example of the sort of question I was referring to. Perhaps this could be:

```
// Note: alternatively this could use -1.0, but it's assumed subtraction would be more costly

File: src/f32ext/ln.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+//excessive precision doesn't matter that much.
+#[allow(clippy::excessive_precision)]
+pub(super) fn ln_1to2_series_approximation(x: f32) -> f32 {
+    // idea from https://stackoverflow.com/a/44232045/
+    // modified to not be restricted to int range and only values of x above 1.0.
+    // and got rid of most of the slow conversions,
+    // should work for all positive values of x.
+    // log base 2(E) == 1/ln(2)
+    //let recip_ln2: f32 = f32::consts::LOG2_E;
```
Comment (tarcieri) [inline src/f32ext/ln.rs]: Why is this line here but commented out?

File: src/f32ext/ln.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    //clippy forces me to do this.
```
Comment (tarcieri) [inline src/f32ext/ln.rs]: It'd be more helpful for the comment here to note clippy's rationale for this

File: src/f32ext/ln.rs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    if abs::abs(x - 1.0_f32) < f32::EPSILON {
+        return 0.0_f32;
+    }
+    let x_less_than_1: bool = x < 1.0;
+    //use fast inv here?
```
Comment (tarcieri) [inline src/f32ext/ln.rs]: Is there a particular fast inverse algorithm you have in mind here?

File: src/f32ext/fract.rs (hunk 1 above)
Comment (Cazadorro) [reply src/f32ext/fract.rs]: The bit manipulation I did at the end of fract can be replaced with a subtraction of `1.0_f32`, I'm not sure whether a subtraction would be worse or better than the bit manipulation I made, particularly in terms of performance where this library would be used.  It appears as if the result would be the same either way. It simplifies the code as well.

File: src/f32ext/fract.rs (hunk 2 above)
Comment (Cazadorro) [reply src/f32ext/fract.rs]: Okay that makes sense! I've never been sure how to phrase those kinds of weird questions not questions

File: src/f32ext/ln.rs (hunk 3 above)
Comment (Cazadorro) [reply src/f32ext/ln.rs]: I had commented it out when compiler warnings complained it wasn't used and I forgot to get rid of it, I believe a similar line exists in exp, and I preemptively put 1/ln(2) in that file thinking I'd need it, and didn't.

File: src/f32ext/ln.rs (hunk 4 above)
Comment (Cazadorro) [reply src/f32ext/ln.rs]: I wasn't used to the rational, it was basically because you can't guarantee direct comparison of floats that would realistically be close enough will be exactly the same number.  I'll go over the code base again to clean up these issues when I get a chance today.

File: src/f32ext/ln.rs (hunk 5 above)
Comment (Cazadorro) [reply src/f32ext/ln.rs]: The inverse approximation that you had already in the code.  At this point I'm not sure it is accurate enough for`ln`, especially when `ln` is compounded in a `powf`, and I'm not sure if it would even be the source of speed loss in a no std environment.

**Review by tarcieri — APPROVED** [review]
Looks good now! Mind also updating the README.md?

**Comment by tarcieri** [comment]
Awesome, thanks!

**Comment by tarcieri** [comment]
Released in v0.4.0 :shipit:
